thunder/distributed/tensor_parallel/row_wise.py

```diff
@@ -0,0 +1,266 @@
+        return clang.slice_in_dim(x, start_idx, start_idx + chunk_size, dim=x.ndim - 1), None
+
+    def postprocess(self, y: TensorProxy, _: Any) -> TensorProxy:
+        # gather `y` along the last dimension
+        import thunder.torch as ltorch
+        from thunder.distributed import prims as dist_prims
+
+        all_reduced = dist_prims.synchronize_tensor_parallel_output(
+            y,
+            self.process_group,
+            RowParallelLinearPrePostProcess.layer_type,
+        )
+        if (bias := self.bias_or_none) is not None:
+            return ltorch.add(all_reduced, bias)
+        else:
+            return all_reduced
+
+    def maybe_modify_args_and_kwargs(self, bsym: BoundSymbol) -> BoundSymbol:
+        """Replace `bias` of `bsym` with `None` if it's Tensor to avoid redundant accumulation.
+
+        The removed `bias` is added by `postprocess` after all-reduce.
+        The local row-wise parallel linear operation with bias could be
+            y_<rank> = linear(x_<rank>, weight_<rank>, bias)
+        which leads to a wrong result after all_reduce as `bias` is added <world_size> times.
+            y = all_reduce(y_<rank>)
+        """
+        if self.bias_or_none is not None:
+            return bsym.from_bsym_swap_proxies({variableify(self.bias_or_none): None}, skip_output=True)
+        else:
+            return super().maybe_modify_args_and_kwargs(bsym)
+
+
+@dataclass
+class RowParallelEmbeddingPreProcess(PrePostProcessInterface):
+    process_group: ProcessGroup
+
+    layer_type: ClassVar[TensorParallelLayerType] = TensorParallelLayerType.ROW_PARALLEL_EMBED
+
+    def preprocess(self, x: TensorProxy) -> tuple[TensorProxy, tuple[Any, ...]]:
+        return super().preprocess(x)
+
+    def postprocess(self, y: TensorProxy, _: Any) -> TensorProxy:
+        from thunder.distributed import prims as dist_prims
+
+        return dist_prims.synchronize_tensor_parallel_output(
+            y,
+            self.process_group,
+            RowParallelEmbeddingPreProcess.layer_type,
+        )
+
+
+@dataclass
+class TransformForRowWiseParallel(TransformForTensorParallel):
+
+    def _calc_new_shape(self, orig_shape: list[int]) -> tuple[int, ...]:
+        new_shape = orig_shape[:]
+        new_shape[1] //= self.process_group.size()
+        return tuple(new_shape)
+
+    def get_visitor_of_computation_trc_and_provenance(
+        self,
+        computation_trace: TraceCtx,
+    ) -> tuple[Callable[[BoundSymbol], VISIT_TYPE], TraceProvenance | str]:
+        from thunder.core.pytree import tree_flatten
+
+        consumers = utils.consumers(computation_trace)
+        flat_args, _ = tree_flatten((computation_trace.args, computation_trace.kwargs))
+        bsym2prepostprocess: dict[BoundSymbol, PrePostProcessInterface] = {}
+        for proxy in filter(lambda p: isinstance(p, TensorProxy), flat_args):
+            for p_name in self.chunked_param_name2layer_type:
```

Comment:
We could simplify and remove the inner for loop replace it with

```python
    layer_type = self.chunked_param_name2layer_type.get(proxy.name)
    if layer_type is not None:
        ...
        if ...:
            match layer_type:
                case ...
```

Reply:
done in e2a8a08274d9e185401fd5c9844c0208eb6b085d